doc/feature/cppPlugins.md

```diff
@@ -0,0 +1,159 @@
+============
+
+This document specifies a modern API for the in-memory C plugin API for
+lossless byte data compressors. It will not replace the 2D image
+compressors and CPU-GPU transfer plugins.
+
+## Requirements
+
+* Support all features used by current lossless byte compressors
+    * Lossless compression of binary blobs
+    * Multiple compressors per plugin DSO
+* Information provided for each compressor:
+    * Compression speed (relative to RLE compressor)
+    * Compression ratio (relative to uncompressed data)
+* Compressors may produce multiple output chunks
+* Compressor engines are instantiated to allow reuse of internal state
+  and memory allocations
+* Easy to use API for plugin implementers
+
+## Dependency Changes
+
+* New: Using Lunchbox plugin API for loading plugins
+
+## API
+
+    const Plugins& Lunchbox::PluginFactory<..>::getPlugins();
+
+    class PluginInfo
+    {
+    public:
+        std::string name;
+        float ratio; //!< Normalized 0..1 size after compression
+        float speed; //!< Relative speed compared to RLE compressor
+    };
+    typedef std::vector< PluginInfo > PluginInfos;
+
+    class PluginRegistry
+    {
+    public:
+        typedef std::unique_ptr< Compressor > CompressorPtr;
+        CompressorPtr choose();
+        CompressorPtr create( const std::string& name );
+    };
+
+    class Compressor
+    {
+    public:
+        typedef lunchbox::Bufferb Result;
+        typedef std::vector< Result > Results;
+
+        virtual const Results& compress( const void* const data,
+                                         const size_t size ) = 0;
+        virtual const Result& decompress( const Results& input,
+                                          const size_t size ) = 0;
+    protected:
+        Results compressed;
+        Result uncompressed;
+    };
+
+
+## Examples
+
+    CompressorPtr PluginRegistry::choose()
+    {
+        float rating = powf( 1.0f, .3f );
+        Plugin* candidate = nullptr;
+
+        const Plugins& plugins = PluginFactory::getInstance().getPlugins();
+        for( const Plugin* plugin : plugins )
+        {
+            for( const PluginInfo& info : plugin->getInfo( ))
+            {
+                float newRating = powf( info.speed, .3f ) / info.ratio;
+                if( newRating > rating )
+                {
+                    candidate = plugin;
+                    rating = newRating;
+                }
+            }
+        }
+        if( candidate )
+            return candidate.construct()
+        return nullptr;
+    }
+
+Compressor implementation:
+
+    #include "zstd/lib/zstd.h"
+
+    namespace pression
+    {
+    namespace plugin
+    {
+    namespace
+    {
+    static const bool LB_UNUSED _initialized = []()
+    {
+        PluginRegistry::registerEngine( { "pression::CompressorZSTD",
+                                         .47f, .25f },
+                                        CompressorZSTD::newInstance );
```

Comment:
First time this new method appears.

If this lambda function the new recommended way of doing static initialization? Where have you seen this pattern?
Is static needed in this case?